### PR TITLE
Allow SSZ-serialized blocks in `publishBlock`

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Adds the ability to submit SSZ-serialized blocks to https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/publishBlock, as outlined in https://github.com/ethereum/beacon-APIs/pull/205

### CODE REVIEW

File: beacon-chain/rpc/eth/beacon/blocks.go [hunk 1]
```diff
@@ -223,6 +225,77 @@ func (bs *Server) SubmitBlock(ctx context.Context, req *ethpbv2.SignedBeaconBloc
...
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return &emptypb.Empty{}, status.Errorf(codes.Internal, "Could not read eth-consensus-version header")
+	}
+	ver := md.Get("eth-consensus-version")
+	if len(ver) == 0 {
+		return &emptypb.Empty{}, status.Errorf(codes.Internal, "Could not read eth-consensus-version header")
+	}
+
+	switch strings.ToLower(ver[0]) {
```
Comment (rauljordan) [inline beacon-chain/rpc/eth/beacon/blocks.go:248]: It's unfortunate this is yet one more place where the code needs to use switch statements to deal with fork versions. We should avoid this at all costs if possible. One idea here is to use Kasey's versioned unmarshaler as an alternative:

> VersionedUnmarshaler represents the intersection of Configuration (eg mainnet, testnet) and Fork (eg phase0, altair). Using a detected VersionedUnmarshaler, a BeaconState or SignedBeaconBlock can be correctly unmarshaled without the need to hard code a concrete type in paths where only the marshaled bytes, or marshaled bytes and a version, are available

This also works for beacon blocks, which is great

File: beacon-chain/rpc/eth/beacon/blocks.go [hunk 2]
```diff
@@ -223,6 +225,77 @@ func (bs *Server) SubmitBlock(ctx context.Context, req *ethpbv2.SignedBeaconBloc
...
+				status.Errorf(
+					codes.InvalidArgument,
+					"could not unmarshal data into %s block: %v",
+					strings.ToLower(ethpbv2.Version_PHASE0.String()),
+					err,
+				)
+		}
+		if err := bs.submitPhase0Block(ctx, block.Block, block.Signature); err != nil {
+			return nil, err
+		}
```
Comment (rauljordan) [inline beacon-chain/rpc/eth/beacon/blocks.go:262]: These `submitXBlock` functions are also unnecessary, as the versioned unmarshaler from the `detect` package returns a signed beacon block interface, which is pretty much all we need to use the generic `submitBlock` function

File: beacon-chain/rpc/eth/beacon/blocks.go (hunk 1 above)
Comment (rauljordan) [reply beacon-chain/rpc/eth/beacon/blocks.go:248]: You can find this under encoding/ssz/detect

File: beacon-chain/rpc/apimiddleware/custom_handlers.go [hunk 3]
```diff
@@ -112,6 +119,53 @@ func handleGetSSZ(
...
+		apimiddleware.WriteError(w, errJson, nil)
+		return true
+	}
+	prepareCustomHeaders(req)
+	if errJson := preparePostedSSZData(req); errJson != nil {
+		apimiddleware.WriteError(w, errJson, nil)
+		return true
+	}
+
+	grpcResponse, errJson := m.ProxyRequest(req)
```
Comment (rauljordan) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go:143]: Most of this logic below exists elsewhere right? Maybe we can deduplicate a little

File: beacon-chain/rpc/eth/beacon/blocks.go [hunk 4]
```diff
@@ -223,6 +226,45 @@ func (bs *Server) SubmitBlock(ctx context.Context, req *ethpbv2.SignedBeaconBloc
...
+// The provided block must be SSZ-serialized.
+func (bs *Server) SubmitBlockSSZ(ctx context.Context, req *ethpbv2.SignedBeaconBlockSSZContainer) (*emptypb.Empty, error) {
+	ctx, span := trace.StartSpan(ctx, "beacon.SubmitBlock")
+	defer span.End()
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return &emptypb.Empty{}, status.Errorf(codes.Internal, "Could not read eth-consensus-version header")
+	}
+	ver := md.Get("eth-consensus-version")
```
Comment (rauljordan) [inline beacon-chain/rpc/eth/beacon/blocks.go:244]: this magic string can perhaps be extracted to be a package variable

File: beacon-chain/rpc/apimiddleware/custom_handlers.go [hunk 5]
```diff
@@ -142,6 +207,30 @@ func prepareSSZRequestForProxying(
 	return nil
 }
 
+func prepareCustomHeaders(req *http.Request) {
+	ver := req.Header.Get("Eth-Consensus-Version")
+	if ver != "" {
+		req.Header.Del("Eth-Consensus-Version")
+		req.Header.Add("Grpc-metadata-Eth-Consensus-Version", ver)
```
Comment (rauljordan) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go:214]: Let's make these package vars

File: beacon-chain/rpc/apimiddleware/custom_handlers.go (hunk 3 above)
Comment (rkapka) [reply beacon-chain/rpc/apimiddleware/custom_handlers.go:143]: Deduplicating things make it look a bit ugly, I would leave the code as is.

**Review by rauljordan — APPROVED** [review]

**Pull request merged**